src/aioswitcher/device/tools.py

```diff
@@ -165,6 +169,38 @@ def convert_str_to_devicetype(device_type: str) -> DeviceType:
     return DeviceType.MINI
 
 
+def convert_token_to_packet(
+    device_type: DeviceType, token: Union[str, None] = None
+) -> Union[str, None]:
+    """Convert a token to token packet.
+
+    Args:
+        device_type: the type of the device.
+        token: the token of the user sent by Email
+
+    Return:
+        Token packet if token is valid,
+        otherwise empty string or raise error.
+
+    """
+    if not device_type.token_needed:
+        return None
+
+    if not bool(token):
+        raise RuntimeError("a token is needed but missing")
```

Comment:
This check should be done where you call this function, you don't need to call it, if you know the device doesn't need a token.
The rest of the function is very clear, it try to convert a token and raise if the conversion fails. but there is no reason to put into this function a device specific logic (the test if the device needs a token) which also require you to pass the device type here just for that, while I think you will need to do this check again where you use it since there are two different APIs anyhow (one with a token and one without) so when you call the API that needs a token you will do the test again.

Reply:
In addition if you do the check where you intend to call the function after you realized you need a token you don't need to return `None` as you do here, you simply do
```python
    if not bool(token):
        raise RuntimeError("a token is needed but missing")
     
     convert_token_to_packet(...
```
 or if you want to to both checks at once (on the caller)
 
 ```python
    if device_type.token_needed and not bool(token):
        raise RuntimeError("a token is needed but missing")
     
     convert_token_to_packet(...
```

Reply:
I got you but @TomerFi asked to remove the `is_valid_token` function so I thought this is the only place we can do this check.. plus don't you think we should avoid this conversation in every API call? That's why i intended to put this conversation with the check in the device init time

Reply:
I agree with @thecode; it would be better to check if you even need a token before invoking the `convert_token_to_packet.`

But it has nothing to do with removing the `is_valid_token` function. Validity and requirement are two different things. If a token is not required, it doesn't necessarily mean it's invalid, but this was the case for your `is_valid_token` function. Even if you would have changed its name, creating a utility function for evaluating a simple boolean value is a tough sell.

Reply:
Did the changes